dapp/src/components/buySell/ApproveSwap.js

```diff
@@ -0,0 +1,264 @@
+import React, { useState, useEffect } from 'react'
+import { fbt } from 'fbt-runtime'
+import { useStoreState } from 'pullstate'
+import { useWeb3React } from '@web3-react/core'
+import ContractStore from 'stores/ContractStore'
+import analytics from 'utils/analytics'
+import withRpcProvider from 'hoc/withRpcProvider'
+import { ethers } from 'ethers'
+import withIsMobile from 'hoc/withIsMobile'
+
+const ApproveSwap = ({
+  stableCoinToApprove,
+  needsApproval,
+  selectedSwap,
+  swapMetadata,
+  onSwap,
+  allowancesLoaded,
+  onMintingError,
+  formHasErrors,
+  swappingGloballyDisabled,
+  storeTransaction,
+  storeTransactionError,
+  rpcProvider,
+  isMobile,
+}) => {
+  const [coinApproved, setCoinApproved] = useState(false)
+  const [stage, setStage] = useState(coinApproved ? 'done' : 'approve')
+  const [contract, setContract] = useState(null)
+  const [isApproving, setIsApproving] = useState({})
+  const web3react = useWeb3React()
+  const { library, account } = web3react
+
+  useEffect(() => {
+    const approval = !(
+      (!selectedSwap ||
+        formHasErrors ||
+        swappingGloballyDisabled ||
+        !allowancesLoaded ||
+        !needsApproval) &&
+      !coinApproved
+    )
+    ContractStore.update((s) => {
+      s.approvalNeeded = approval
+    })
+  }, [selectedSwap, needsApproval])
```

Comment:
Shouldn't other things also trigger an execution of `useEffect` besides `selectedSwap &needsApproval`?
Things like: formHasErrors, swappingGloballyDisabled, allowancesLoaded, coinApproved?